### DISCUSSION TITLE
Encoder output float or s32 ?

### DISCUSSION THREAD

**Discussion by ghost** [multigcs/LinuxCNC-RIO#13] [Q&A]
As a bodge to get my mpg encoders to work with jogging, I messed around with rio.c to get a s32 output as opposed to float. I can see the vin class would need a float for vin_frequency & vin_sonar. Is there a way other than the bodge I used or using conv_float_s32 ?

Haven't had a chance to play with the spindle encoder, spent the weekend getting a network connection into the garage, and other bits of life.

But in saying that I have got the Z axis moving on the lathe, tho the X aixs still requires work.

**Comment by multigcs** [comment]
i thought linuxcnc doesn't care if it's float or s32,
I will test this with the MPG.
do you have an example configuration for MPG ?

**Comment by ghost** [comment]
This is my hal file file for the mpg. The comments at the top are for my sanity.
```
#  Dual axis lathe pendant
#  in lathe-pendant.0.X-encoder-counts
#  in lathe-pendant.0.Z-encoder-counts
#  out lathe-pendant.0.mpg-X-counts
#  out lathe-pendant.0.mpg-X-enable
#  out lathe-pendant.0.mpg-X-scale
#  out lathe-pendant.0.mpg-Z-counts
#  out lathe-pendant.0.mpg-Z-enable
#  out lathe-pendant.0.mpg-Z-scale
#  in lathe-pendant.0.mpg-enable
#  in lathe-pendant.0.scale-select-1
#  in lathe-pendant.0.scale-select-2
#  in lathe-pendant.0.scale-select-3
#  param lathe-pendant.0.mpg-scale1 0.1000
#  param lathe-pendant.0.mpg-scale2 0.0100
#  param lathe-pendant.0.mpg-scale3 0.0025

# MPG Pendant
# Sumary of rio pins used
# Encoder X rio.PV.00 Board 2 DB-02 DB-03
# Encoder Z rio.PV.01 Board 2 DB-04 DB-05
# MPG Enable Board 2 DB-06 ENABLE MPG	rio.din.08-not
# Board 2 DB-07 MPG-SELECT 0.100		rio.din.05-not
# Board 2 DB-08 MPG-SELECT 0.010		rio.din.06-not
# Board 2 DB-09 MPG-SELECT 0.001		rio.din.07-not

# Load comp & add to servo thread
loadrt lathe_pendant
addf lathe-pendant.0 servo-thread

# Setup scale values
setp lathe-pendant.0.mpg-scale1 1.000
setp lathe-pendant.0.mpg-scale2 0.100
setp lathe-pendant.0.mpg-scale3 0.010

# Connect component inputs
net x_counts_in rio.PV.00
net x_counts_in lathe-pendant.0.X-encoder-counts

net z_counts_in rio.PV.01
net z_counts_in lathe-pendant.0.Z-encoder-counts

net enable_in rio.din.08-not
net enable_in lathe-pendant.0.mpg-enable

net scale_select_1_in rio.din.05-not
net scale_select_1_in lathe-pendant.0.scale-select-1
net scale_select_2_in rio.din.06-not
net scale_select_2_in lathe-pendant.0.scale-select-2
net scale_select_3_in rio.din.07-not
net scale_select_3_in lathe-pendant.0.scale-select-3




# 
net x_enable_out lathe-pendant.0.mpg-X-enable
net z_enable_out lathe-pendant.0.mpg-Z-enable
net x_scale_out lathe-pendant.0.mpg-X-scale
net z_scale_out lathe-pendant.0.mpg-Z-scale
net x_counts_out lathe-pendant.0.mpg-X-counts
net z_counts_out lathe-pendant.0.mpg-Z-counts

# Enables
# Axis Enable
net x_enable_out axis.x.jog-enable
net z_enable_out axis.z.jog-enable
# Joint Enable
net x_enable_out joint.0.jog-enable 
net z_enable_out joint.1.jog-enable

# Scales
# Axis Scale
net x_scale_out axis.x.jog-scale
net z_scale_out axis.z.jog-scale
# Joint Scale
net x_scale_out joint.0.jog-scale
net z_scale_out joint.1.jog-scale


# Counts
# Axis Counts
net x_counts_out axis.x.jog-counts
net z_counts_out axis.z.jog-counts
# Joint Counts
net x_counts_out joint.0.jog-counts
net z_counts_out joint.1.jog-counts
```

And the lathe-pendant component, I use a variation of this on my mill as well. It does nothing with the MPG counts in this version, it's main function is to set the scale for the joints\axis based on 3 switches. I could have done this in hal as well but I find a quick component is portable and less messy in the hal file.

Just run:
sudo halcompile --install lathe_pendant.comp

```
component lathe_pendant "2 Axis Lathe Pendnant";
description "2 Axis (2 of MPG & 1 of DP4T switch) pendant inspired by the mpg component of Kurt Jacobson.";


//MPG Pendant
// Sumary of rio pins used
// Encoder X rio.PV.00 Board 2 DB-02 DB-03
// Encoder Z rio.PV.01 Board 2 DB-04 DB-05
// MPG Enable Board 2 DB-06 ENABLE MPG	rio.din.06
// Board 2 DB-07 MPG-SELECT 0.100		rio.din.07
// Board 2 DB-08 MPG-SELECT 0.010		rio.din.08
// Board 2 DB-09 MPG-SELECT 0.001		rio.din.09

pin in bit mpg_enable;

pin in bit scale_select_1;
pin in bit scale_select_2;
pin in bit scale_select_3; 

pin in s32 X_encoder_counts;
pin in s32 Z_encoder_counts;

param rw float mpg_scale1=0.100;
param rw float mpg_scale2=0.010;
param rw float mpg_scale3=0.0025;

//Enable;
pin out bit mpg_X_enable;
pin out bit mpg_Z_enable;

//MPG Scale
pin out float mpg_X_scale;
pin out float mpg_Z_scale;

//Encoder Counts;
pin out s32 mpg_X_counts;
pin out s32 mpg_Z_counts;

function _;
license "GPL"; // indicates GPL v2 or later
;;

FUNCTION(_){

	if ( mpg_enable) {	
	
		mpg_X_enable = 1;
		mpg_Z_enable = 1;

				
		if ( scale_select_3 && !scale_select_2 && !scale_select_1 ) {
				mpg_X_scale = mpg_scale3; 
				mpg_Z_scale = mpg_scale3;
				}
			else if ( !scale_select_3 && scale_select_2 && !scale_select_1  ) {
				mpg_X_scale = mpg_scale2; 
				mpg_Z_scale = mpg_scale2;
				}		
			else if ( !scale_select_3 && !scale_select_2 && scale_select_1  ) {
				mpg_X_scale = mpg_scale1; 
				mpg_Z_scale = mpg_scale1;
				}
// 2 or more select bits high
			else {
				//mpg_X_scale = 0; 
				//mpg_Z_scale = 0;
				}	
			}	

// Enable bit low
	else {
			mpg_X_enable = 0;
			mpg_Z_enable = 0;
		}
	
	mpg_X_counts = X_encoder_counts;
	mpg_Z_counts = Z_encoder_counts;
}

```

Looking at http://linuxcnc.org/docs/2.9/html/man/man9/motion.9.html 
axis.L.jog-counts & joint.N.jog-counts both require a S32 input.

When I was trying to get the vin_mpgencode to work I had to use `output signed [BITS-1:0] pos` to get it to work as I wanted. I don't know if this is a defect on my part, hahaha.

Note that I changed DIN & DOUT & PV to have leading zeroes, just a personal thing for me.

The Linuxcnc docs have a simple MPG example for use with Parallel Port, http://linuxcnc.org/docs/2.9/html/examples/mpg.html , but doesn't include the HAL connections for using joints which allows you to jog prior to homing.

**Comment by multigcs** [comment]
i have add new signals with trailing -s32

so:
 rio.PV.0 is float

 rio.PV.0-s32 is S32

hope this helps, i can now use an encoder as jog on a mill

**Comment by jschoch** [comment]
@ozzyrob i'm thinking about doing a wireless pendant, any thoughts on how to go about it? 
 What pendant are you using?
How well does it work with the rio setup?  
What is that HAL component you have in your config above?

**Comment by ghost** [comment]
Personally I use a wired pendant for the mill connected to a MESA setup (I wouldn't consider using a wireless pendant, I'd put it down and gremlins in my workshop would move it).

For the lathe I'm using hostmot2 firmware, connected via EPP, on a Spartan6 dev card with some custom PCB's.

The hal component just puts most of the HAL routing into a single component. As per my description " It does nothing with the MPG counts in this version, it's main function is to set the scale for the joints\axis based on 3 switches. I could have done this in hal as well but I find a quick component is portable and less messy in the hal file."

**Comment by jschoch** [comment]
those gremlins are also in my shop...
